### PR TITLE
Azure: validate UltraSSD instance types in multi-zone regions

### PR DESCRIPTION
**Pull request by r4f4**
In order to validate UltraSSD in multi-zone regions, I have updated the Azure client to return `LocationInfo` information which we then use to check if availability zones are supported and which ones have UltraSSD available.

The error messages might still need some more polish before this is merged but I would like to get some comments, specially on 58ad37dbd400302d17c4faf121e8d13db8d99c47, before I proceed ironing out the details.

I have also added unit tests for as many combinations as I found relevant but they are by no means exhaustive.

### CODE REVIEW

**Comment by r4f4** [comment]
/hold
Until I get feedback on 79d6587020d03c8c91a38017cb70b1fb489115cd

File: pkg/asset/installconfig/azure/client.go [hunk 1]
```diff
@@ -248,12 +249,17 @@ func (c *Client) GetVirtualMachineSku(ctx context.Context, name, region string)
 			if !strings.EqualFold(name, *sku.Name) {
 				continue
 			}
+			return &sku, nil
+
+			// FIXME: I don't think the following block is needed anymore since
```
Comment (rna-afk) [inline pkg/asset/installconfig/azure/client.go:254]: I don't think we need it if you are confident the filter works. Haven't tried it myself but it looks like it should work.

**Review by rna-afk — COMMENTED** [review]
LGTM. Can't see any complications coming from switching to the service but will wait for others' opinion

File: pkg/asset/installconfig/azure/client.go (hunk 1 above)
Comment (r4f4) [reply pkg/asset/installconfig/azure/client.go:254]: Right. I'll remove it if I don't see anything suspicious in the CI runs.

File: pkg/asset/installconfig/azure/validation.go [hunk 2]
```diff
@@ -127,7 +129,7 @@ func ValidateInstanceType(client API, fieldPath *field.Path, region, instanceTyp
 				errMsg := fmt.Sprintf("PremiumIO not supported for instance type %s", instanceType)
 				allErrs = append(allErrs, field.Invalid(fieldPath.Child("osDisk", "diskType"), diskType, errMsg))
 			}
-		} else if strings.EqualFold(*capability.Name, "UltraSSDAvailable") {
+		} else if strings.EqualFold(*capability.Name, ultraSSDCapability) {
```
Comment (patrickdillon) [inline pkg/asset/installconfig/azure/validation.go:132]:
```suggestion
		} else if strings.EqualFold(*capability.Name, aztypes.AcceleratedNetworkingEnabled) {
```
This should be fixed by either #5840 or more probably #5846, although of course you are free to fix here.

File: pkg/asset/installconfig/azure/validation.go [hunk 3]
```diff
@@ -137,6 +139,37 @@ func ValidateInstanceType(client API, fieldPath *field.Path, region, instanceTyp
 		}
 	}
 
+	// Convenience check so the tests don't have to define LocationInfo if it's not used
+	if typeMeta.LocationInfo != nil && ultraSSDEnabled {
+		for _, locationInfo := range *typeMeta.LocationInfo {
+			// If Availability Zones are not supported (e.g StackCloud)
+			if locationInfo.ZoneDetails == nil {
+				continue
+			}
```
Comment (patrickdillon) [inline pkg/asset/installconfig/azure/validation.go:148]: This case is already handled by the preceding conditional two lines earlier.

**Review by patrickdillon — COMMENTED** [review]
This looks great. A couple of suggestions.

File: pkg/asset/installconfig/azure/validation.go (hunk 3 above)
Comment (r4f4) [reply pkg/asset/installconfig/azure/validation.go:148]: In my tests against the API for the Azure Stack case the `LocationInfo` struct was defined but the `ZoneDetails` field was `nil`. So I need to check that before dereferencing the pointer in the `for` loop that follows. I even added that as a test case https://github.com/openshift/installer/pull/5854/files#diff-310a64406bdb016820cd032ef2ca5cd092ba942f2b7ccbedeec0abc254fbef74R670-R676

File: pkg/asset/installconfig/azure/validation.go (hunk 2 above)
Comment (r4f4) [reply pkg/asset/installconfig/azure/validation.go:132]: You mean I should add `UltraSSDAvailable` as a type in `aztypes`?

**Comment by r4f4** [comment]
I've also implemented an alternative approach in https://github.com/r4f4/installer/tree/azure-ultrassd-alt

**Comment by Amoghrd** [comment]
Pre-merge testing works fine in all regions with appropriate error messages.

**Comment by r4f4** [comment]
/close
Closing this in favour of https://github.com/openshift/installer/pull/5937

**Pull request closed without merging**